Das 137

mads-bertelsen opened the pull request:
Upload of a Jupyter Notebook that demos the capabilities of the dream module. It was uploaded on request of Simon Heybrock. The code relies on recent changes to the dream module, so the other active pull requests should be considered first.

The notebook was cleared before uploading as instructed.

Commit:
Removed test code from dream.py and moved conversion to dspacing into a separate module.

Commit:
Removed local import of scipp module left in by accident.

Commit:
Updates to dream.py in connection with review of Jupyter Notebook using the code performed by Celine.

A bad variable names were improved: d -> data
Use scipy.constants were possible instead of inline constant definitions
Fixed a few typos in docstrings
Added some missing units in docstrings
Removed _ from _generate_random_sample method, as users may use it, and it is sufficiently documented

Commit:
Updated dependencies and typos

SimonHeybrock commented:
> The notebook was cleared before uploading as instructed.

@mads-bertelsen The notebook in this branch *does* contain outputs. Can you double check that you clear them before saving?

Commit:
Cleared notebook.

mads-bertelsen commented:
@SimonHeybrock I don't understand how that happened. I see it wasn't cleared, even in my local version, but clearing it again in exactly the same way worked. Sorry about that, good catch!

SimonHeybrock commented on dream_demo.ipynb:
```diff
@@ -0,0 +1,261 @@
+    "n_pixel = int(2E4)\n",
+    "n_events = int(2E6)\n",
+    "\n",
+    "Dream = dream.DreamTest(n_pixel) # Create a fake dream powder diffractometer with detector positions and calibration data\n",
```
Uppercase names are usually only used for classes, not for variables (see https://www.python.org/dev/peps/pep-0008/). Reading `Dream`, most people would think it is a class, but it is not.

mads-bertelsen replied:
The issue arises because the name can not be the same as the module name, as the cell in the notebook can not be rerun when that is the case.

SimonHeybrock replied:
You will need to find a different name, or import the module under a different name (`import dream as xxx`). As soon as some `autopep8` check such as `flake8` (which we have in the `scipp` repository) is introduced, this will fail. Also, we should not teach bad style to users.

SimonHeybrock commented on dream_demo.ipynb:
```diff
@@ -0,0 +1,261 @@
+   "outputs": [],
+   "source": [
+    "dspacing = powder_reduction.powder_reduction(Dream.data, calibration=Dream.cal)\n",
+    "print(dspacing) # Show the converted dataset"
```
Do you feel that comments like this are necessary in a notebook? Isn't it self documenting, given that the output will appear immediately below?

mads-bertelsen replied:
It was reviewed internally without those comments, and more explanations were requested.

SimonHeybrock commented on dream_demo.ipynb:
```diff
@@ -0,0 +1,261 @@
+   "metadata": {},
+   "source": [
+    "## Plotting the dspacing dataset\n",
+    "First the raw data is collected from each of the sparse channels in the detector array. Next `numpy` is used to histogram the data, after which the data is plotted with `matplotlib`. We see a diffraction pattern with peaks in the specified interval."
```
Why are you using `numpy` for histogramming instead of `scipp`?

mads-bertelsen replied:
Plotting is now done using scipp histogramming and scipp plotting.

SimonHeybrock commented on dream_demo.ipynb:
```diff
@@ -0,0 +1,261 @@
+    "\n",
+    "data_to_plot = sc.Dataset()\n",
+    "N = 501\n",
+    "bins = np.linspace(2.0, 10.0, N).astype(np.float64)\n",
```
Since yesterday, we can plot sparse data directly (binning internally), see https://scipp.readthedocs.io/en/latest/user-guide/plotting.html#Plotting-sparse-data.

Commit:
Updated dtype when using 3D positions in array.

Commit:
Update of Dream.py to remove unnecessary dense Tof DataArray that was always generated.
This was required due to a bug in conversion that has since been fixed.

The Dream_demo notebook was updated to use scipp features for plotting and histogramming.

Commit:
Renamed improperly named variabled Dream which was an object instance, but the capitalization implied it was a class.
It is now named instrument.

mads-bertelsen commented:
The dream.py module has been modified as well, the unnecessary dense Tof coords were removed. This can be done as a bug that previously required such dummy coords to be present was fixed.

SimonHeybrock commented on dream_demo.ipynb:
```diff
@@ -0,0 +1,268 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%matplotlib inline\n",
```
This should not be needed. `scipp.plot` should setup things correctly. It works for me without, does it not for you?

mads-bertelsen replied:
The %matplotlib inline was requested by a reviewer who couldn't run without it, I did not have issues. I suggest we leave it for wider compatibility.

SimonHeybrock replied:
I suggest we track down the issue. How did the reviewer run it?
